boot/bootutil/src/boot_record.c

```diff
@@ -0,0 +1,228 @@
+                      const struct image_header *hdr,
+                      const struct flash_area *fap)
+{
+
+    struct image_tlv_iter it;
+    uint32_t offset;
+    uint16_t len;
+    uint16_t type;
+    size_t record_len = 0;
+    uint8_t image_hash[32]; /* SHA256 - 32 Bytes */
+    uint8_t buf[MAX_BOOT_RECORD_SZ];
+    uint32_t boot_record_found = 0;
+    uint32_t hash_found = 0;
+    uint16_t ias_minor;
+    int32_t res;
+    enum shared_memory_err_t res2;
+
+    /* Manifest data is concatenated to the end of the image.
+     * It is encoded in TLV format.
+     */
+
+    res = bootutil_tlv_iter_begin(&it, hdr, fap, IMAGE_TLV_ANY, false);
+    if (res) {
+        return BOOT_STATUS_ERROR;
+    }
+
+    /* Traverse through the TLV area to find the boot record
+     * and image hash TLVs.
+     */
+    while (true) {
+        res = bootutil_tlv_iter_next(&it, &offset, &len, &type);
+        if (res < 0) {
+            return BOOT_STATUS_ERROR;
+        } else if (res > 0) {
+            break;
+        }
+
+        if (type == IMAGE_TLV_BOOT_RECORD) {
+            if (len > sizeof(buf)) {
+                return BOOT_STATUS_ERROR;
+            }
+            res = flash_area_read(fap, offset, buf, len);
+            if (res) {
+                return BOOT_STATUS_ERROR;
+            }
+
+            record_len = len;
+            boot_record_found = 1;
+
+        } else if (type == IMAGE_TLV_SHA256) {
+            /* Get the image's hash value from the manifest section. */
+            if (len > sizeof(image_hash)) {
+                return BOOT_STATUS_ERROR;
+            }
+            res = flash_area_read(fap, offset, image_hash, len);
+            if (res) {
+                return BOOT_STATUS_ERROR;
+            }
+
+            hash_found = 1;
+
+            /* The boot record TLV is part of the protected TLV area which is
+             * located before the other parts of the TLV area (including the
+             * image hash) so at this point it is okay to break the loop
+             * as the boot record TLV should have already been found.
+             */
+            break;
+        }
+    }
+
+
+    if (!boot_record_found || !hash_found) {
+        return BOOT_STATUS_ERROR;
+    }
+
+    /* Update the measurement value (hash of the image) data item in the
+     * boot record. It is always the last item in the structure to make
+     * it easy to calculate its position.
+     * The image hash is computed over the image header, the image itself and
+     * the protected TLV area (which should already include the image hash as
+     * part of the boot record TLV). For this reason this field has been
+     * filled with zeros during the image signing process.
+     */
+    offset = record_len - sizeof(image_hash);
+    /* Avoid buffer overflow. */
+    if ((offset + sizeof(image_hash)) > sizeof(buf)) {
+        return BOOT_STATUS_ERROR;
+    }
+    memcpy(buf + offset, image_hash, sizeof(image_hash));
```

Comment:
Not sure why you are reading the sha256 TLV, appending it here, and then seem to be passing the length of the buffer without the added hash to `boot_add_data_to_shared_area`, which itself doesn't seem to be using the hash at all; what am I missing?

Reply:
The hash is not appended the boot record, it overwrites the last 32 bytes (sizeof(image_hash)) of the boot record that is in the buffer ('buf' is pointing to the first element of this array, so this memcpy is writing 32 bytes from this element: buf[record_len - sizeof(image_hash)]).

Reply:
Ah got it, so you are adding 32 byte `\0` string and overriding it here from the read sha256 TLV! :+1:

Reply:
Yes.